### PR TITLE
μ-attention eval suite + methodology report — what a learned layer adds over frozen e5

### PR DESCRIPTION
**Pull request by s243a**
This branch builds the retrieval/applications + evaluation layer for the **μ-attention** model (directional fuzzy
membership `μ(node|root)` over **frozen** e5) and rigorously answers one question: **μ sits on top of frozen e5 —
what does the learned layer add over raw e5 cosine?** It surfaces honest negatives alongside the wins.

**📋 Review entry point: [`prototypes/mu_cosine/REPORT_eval_methodology.md`](prototypes/mu_cosine/REPORT_eval_methodology.md)** —
self-contained (model, baseline, datasets, metrics, all result tables, threats-to-validity, and open questions).
Deeper theory/results in `DESIGN_model_applications.md`; data-acquisition strategy in `DESIGN_wikipedia_sampling.md`.

### Core finding
μ does **not** beat e5 as a *symmetric ranker* (tested 5 ways — e5 is competitive-or-better on coarse/medium
relatedness and high-recall thresholding). μ wins on the axes a frozen symmetric encoder **structurally lacks**:
- **Directionality** — `AUC(μ(member|container) > μ(container|member))` **0.78 vs e5's 0.51 coin-flip** (e5 asymmetry 0.001).
- **Close-negative discrimination** — true-parent vs *sibling*: μ **0.73 vs e5 0.62**, and μ's edge **grows as
  negatives get closer** (e5 scores parent 0.832 ≈ sibling 0.815 — can't tell member-of from sibling-of).
- **Calibrated degrees** — μ has 4× the dynamic range; e5 squashes everything into a 0.76–0.84 band (can't express "low").
- **Data scaling** — on a real OOD task (Pearltrees bookmark filing), a 3-seed fine-tune **crosses the e5-cos bar
  (MRR 0.291) at ~30% of the data and is still rising at 100%** (0.358; recall@10 0.573 vs 0.440).

### The e5 calibration issue (central methodological point)
e5-small cosines are compressed (~0.76–0.84), so a parent, a cross-subdomain category, and a *sibling* all score
~0.8. Rank-AUC hides this; the usable signal is the **calibrated degree** (μ=0.22 ⇒ "barely a member"), not a
threshold. e5 is **frozen** (at its ceiling), μ is the **trained** layer (data-scalable) — so in-domain data
improves μ, not e5; the learning curve is the evidence.

### What works / what doesn't
- **Works:** directionality, close-neighbour disambiguation, calibrated degrees, data-scaling (filing curve).
- **Doesn't:** μ as a symmetric rank metric (e5 ≥ μ on easy/medium negatives); calibration ≠ clean high-recall
  thresholding (FPR leaky); clean-domain coverage gave no rank gain (e5 already strong there); zero-shot OOD μ loses.
- **Implied architecture:** a **hybrid** — e5 for coarse symmetric ranking, μ for the decisive top-of-list cases
  (direction + close-neighbour + calibrated degree). Not "μ replaces e5."

### Methodology notes (surfaced for fair review)
We document our own missteps: the first filing eval (member→*exact*-parent vs all folders) was a misaligned
classification task that favoured e5; an early "hard negative" wasn't close enough (siblings are the real test).
Single-run discrimination evals are n=400–2000 (e5-cos wobbled ±0.013 at n=400); the filing curve is 3-seed.

### Tooling
`eval_filing.py` (filing/home-turf/node-holdout/`--drop-admin`), `eval_relatedness.py` (pair discrimination +
directional + FPR), `train_filing.py` (multi-seed fine-tune), `train_mu_attention.py --eval-retrieval`. Harvested
data and `*.pt` checkpoints are gitignored build outputs.

### For the reviewer (REPORT §9)
1. Is `AUC(fwd>rev)` the right way to credit directionality, or is a hierarchy-reconstruction task stronger?
2. Should calibration be reported via a proper metric (ECE / reliability diagram / d-prime) rather than per-stratum means?
3. Is "sibling" the right hardest negative, and the easy→hard→sibling *trend* the right way to show e5's degradation?
4. Is the frozen-e5-ceiling vs trained-μ-headroom data-scaling argument sound — what would make the extrapolation rigorous (more fractions, fitted curve + CI, a 2nd OOD domain)?
5. Does the evidence support the **hybrid** conclusion, or μ-alone for the directional use case?

### CODE REVIEW

File: prototypes/mu_cosine/REPORT_eval_methodology.md [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+### 4.4 Directionality — μ's decisive, structural win
+For 1,500 membership edges: `μ(member|container)` (forward) should be high, `μ(container|member)` (reverse) low.
+
+| scorer | mean fwd | mean rev | asymmetry | AUC(fwd > rev) |
+|---|---|---|---|---|
+| **mu-elem** | 0.512 | 0.185 | **0.326** | **0.776** |
+| e5-cos | 0.838 | 0.837 | 0.001 | 0.509 |
+
+**e5-cos cannot tell which way membership points (AUC 0.509 — a coin flip);** μ gets it right 78% of the time.
+The `query:`/`passage:` prefix gives e5 essentially nothing (asymmetry 0.001). This is the property a symmetric
```
Comment (s243a) [inline prototypes/mu_cosine/REPORT_eval_methodology.md:121]: **[blocking] Directionality headline is not reproducible from committed tooling**

`AUC(fwd>rev)` is a good necessary check for asymmetry, but the committed `eval_relatedness.py` appears to compute only symmetric POS/HARD/EASY discrimination and FPR. I could not find a committed path that produces this forward-vs-reverse table. Please either commit the directional eval script/command and seed/split inputs, or mark this number as derived from an external/manual analysis. Also make the test explicitly paired per edge: score `μ(child|parent)` and `μ(parent|child)` on the same relation, then bootstrap the paired AUC/accuracy CI.

File: prototypes/mu_cosine/REPORT_eval_methodology.md [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+metric *cannot* have.
+
+### 4.5 Close-negative discrimination — μ's advantage grows as negatives get closer
+The hardest negative is a **sibling** (another child of the same parent — same fine topic, *no* membership
+relation, often *more* e5-similar than the true parent). 2,000 (child, parent, sibling) triples.
```
Comment (s243a) [inline prototypes/mu_cosine/REPORT_eval_methodology.md:126]: **[blocking] Sibling close-negative result also needs reproducible generation and false-negative filtering**

The sibling stratum is the right hardest negative for this claim, but it is also where false negatives are most likely in a multi-inheritance/transitive graph. Please commit the `(child, parent, sibling)` generation/eval path and document filters excluding direct parents, transitive ancestors, aliases/redirects, and valid multi-parent categories. Without that, the 0.73 vs 0.62 result is hard to audit and may partly reflect negative-label noise.

File: prototypes/mu_cosine/REPORT_eval_methodology.md [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+(different coarse).
+
+| scorer | AUC(POS vs HARD) | AUC(POS vs EASY) | mean POS / HARD / EASY |
+|---|---|---|---|
+| e5-cos | **0.726** | 0.838 | 0.814 / 0.781 / 0.761 |
+| mu-super | 0.671 | 0.814 | **0.367 / 0.214 / 0.095** |
+
+**On rank-AUC, e5 is competitive-or-better.** But note the means: **e5 squashes all strata into a 0.05 band
+(0.76–0.81); μ has a 4× dynamic range (0.37/0.21/0.10)** — the calibration point (§5).
+
```
Comment (s243a) [inline prototypes/mu_cosine/REPORT_eval_methodology.md:101]: **[blocking] Per-stratum means show dynamic range/separation, not calibration**

This table supports “μ has a wider dynamic range” and maybe better stratum separation, but not calibration in the probabilistic sense. A statement like `μ=0.22` meaning “barely a member” needs a reliability check against empirical membership rates, or at least d-prime / KS / score-distribution overlap if the claim is separability. Please either rename this throughout to “dynamic range / degree readability” or add ECE with equal-frequency bins, reliability diagrams, Brier/log score after an explicit calibration map, and d-prime per stratum.

File: prototypes/mu_cosine/REPORT_eval_methodology.md [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+training seeds, fixed eval split**, ✓ = mean−sd above the e5-cos bar 0.291):
+
+| data frac | n_train | MRR (mean±sd) | recall@10 | vs e5-cos |
+|---|---|---|---|---|
+| 0.10 | 492 | 0.230 ± 0.025 | 0.406 | −0.061 |
+| 0.30 | 1478 | 0.317 ± 0.006 | 0.537 | **+0.025 ✓** |
+| 1.00 | 4929 | 0.358 ± 0.018 | 0.573 | **+0.066 ✓** |
+
+μ **crosses the e5-cos bar between 10–30% of the data and keeps rising** (no plateau); at 100%, recall@10 0.573
+vs 0.440, median rank 6 vs 17. The cross survives seed noise (mean−sd clears the bar at 30% and 100%).
```
Comment (s243a) [inline prototypes/mu_cosine/REPORT_eval_methodology.md:75]: **[high] The learning-curve claim is directionally interesting but statistically underpowered**

Three seeds and three-ish data fractions are enough for a promising pilot, but not enough for “keeps rising / no plateau” or a precise crossing point. Please report confidence intervals at each fraction, use paired splits, add more data fractions, and fit a saturating/power-law curve with a CI on the crossing point. The current curve can support “crossed e5 on this split by 30% and 100%,” but not a robust extrapolation claim.

File: prototypes/mu_cosine/REPORT_eval_methodology.md [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+
+- **μ-attention** produces `μ(node | root) ∈ [0,1]` — a **directional, fuzzy** membership/relatedness degree over
+  a concept graph. It is a small permutation-invariant transformer over **frozen** e5-small-v2 embeddings (the
+  root uses e5's `query:` role, candidates/ancestors use `passage:`, which is what lets `μ(a|b) ≠ μ(b|a)`), read
+  out per **operator** (`SYM` lateral; `WIKI`/`ELEM` directional membership). Only the attention/readout/operator
+  params are trained; **e5 is frozen.** `mu-super` = equal-weight operator superposition; `mu-elem` = the
+  `element_of` operator alone.
+- **Baseline: `e5-cos`** = cosine of the frozen e5 vectors (`query:` node vs `passage:` root). Zero trained
+  parameters. Because μ is built *on* e5, e5-cos is the honest "does the learned layer earn its keep" baseline.
+- **Key asymmetry of the comparison:** e5 is **frozen** (fixed at pretraining — no in-domain data can improve
```
Comment (s243a) [inline prototypes/mu_cosine/REPORT_eval_methodology.md:21]: **[high] Frozen e5 is a fair product baseline, but not an architecture-control baseline**

`e5-cos` is the right baseline for “does μ add value over raw frozen e5,” but the data-scaling section compares a trained μ head against an untrained frozen score. That confounds μ-specific architecture with generic in-domain adaptation. Please add, or list as a required follow-up, a frozen-e5 linear probe / logistic calibration / temperature or isotonic rescale trained on the same data, and ideally LoRA/full e5 fine-tune as upper-bound controls.

**Review by s243a — COMMENTED** [review]
Council review summary: **request changes** on methodology/reproducibility. Opus 4.8, GPT-5.5, Gemini, and the deep-research pass all agreed the report is unusually honest and the **hybrid** conclusion is probably the right interpretation: e5 remains the strong symmetric ranker, while μ is valuable for directionality, close-neighbour disambiguation, and readable membership degrees.

The blockers are claim-vs-evidence gaps in a methodology report:
- The decisive structural-win tables for directionality and sibling close-negatives are not reproducible from committed tooling as currently described.
- The report uses “calibration” for per-stratum means / dynamic range, but does not measure calibration in the probabilistic sense.
- The data-scaling claim needs stronger statistics before saying “keeps rising” or implying architecture headroom rather than generic in-domain adaptation.

Suggested path to merge: commit or document the exact scripts for §4.4/§4.5, rename or measure calibration, add CIs/seeds for headline AUCs, add a sibling false-negative check, and include at least one trained-on-frozen-e5 control such as a linear probe or temperature/isotonic calibration baseline.

File: prototypes/mu_cosine/REPORT_eval_methodology.md [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+- μ is **not** a better *symmetric rank* metric than e5 on easy/medium negatives (e5 AUC ≥ μ there).
+- μ's calibration does **not** give clean high-recall thresholding (FPR leaky; distributions overlap).
+- **Clean-domain coverage gave no rank gain** (e5 already strong on clean content).
+- Zero-shot OOD μ loses badly (it needs in-domain data).
+
+**Architecture implied:** a **hybrid** — e5 for cheap coarse symmetric ranking; **μ for direction + close-
+neighbour disambiguation + calibrated degree** (the practically decisive top-of-list cases). Not "μ replaces e5."
+
+## 8. Threats to validity / limitations (for the reviewer)
+
```
Comment (s243a) [inline prototypes/mu_cosine/REPORT_eval_methodology.md:285]: **[medium] Threats-to-validity should include the missing baselines and reproducibility gaps**

This section is refreshingly candid, but it should explicitly add: unreproducible directional/sibling scripts until committed; sibling false-negative risk; in-sample threshold selection for FPR@90%TPR unless thresholds are chosen on a validation split; missing trained-head-on-frozen-e5 controls; and CI/seed limitations for the headline single-run AUCs. Those are the main threats reviewers will ask about.

File: prototypes/mu_cosine/REPORT_eval_methodology.md [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+### 4.1 OOD filing (real labels) — zero-shot, then a data-scaling curve
+Rank candidate folders by score; ground truth = the bookmark's actual folder (335 folders ≥3 bookmarks, 500
+sampled queries).
+
+**Zero-shot** (`model_nodetype`, never trained on filing):
+
+| ranker | recall@1 | recall@10 | MRR |
+|---|---|---|---|
+| **e5-cos** | **0.202** | **0.480** | **0.299** |
+| mu-super | 0.088 | 0.248 | 0.151 |
```
Comment (s243a) [inline prototypes/mu_cosine/REPORT_eval_methodology.md:57]: **[low] e5 baseline number drifts across report sections**

This table reports e5-cos MRR as `0.299`, while the later learning-curve table uses an e5-cos bar of `0.291` and different recall@10 values. If these are different filtered/eval splits, please label them explicitly; otherwise normalize the baseline so the crossing claim is not comparing against a moving target.

File: prototypes/mu_cosine/REPORT_eval_methodology.md [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+  3-seed, §4.6).** μ-architectural after all.
+- **Close-negative (sibling):** same — symmetric-trained μ lost, discriminatively-trained **μ beats the probe
+  (0.812 vs 0.777, node-disjoint 3-seed).**
+- **Symmetric rank:** e5 ≥ μ (a genuine non-win — μ's edge is direction + close-neg + readable degree, not coarse rank).
+- **Calibrated degrees:** μ has 4× dynamic range, but that is *readability/separation*, not probabilistic
+  calibration, and does not give clean high-recall thresholding (FPR leaky). Rename pending (§5).
+- **Clean-domain coverage:** no rank gain (e5 already strong).
+- **Data scaling:** fine-tuning crossed the **e5-*cos*** bar — but that baseline is uncontrolled; an e5-*probe*
+  curve is the needed comparison and is **not yet run**.
+
```
Comment (s243a) [inline prototypes/mu_cosine/REPORT_eval_methodology.md:271]: **[nit] Stale filing-baseline wording**

§6 now reports the filing e5-probe as measured and weak, but this line still says the needed comparison is “not yet run.” Please reconcile §7/§8 with §6 so the report has one status for the filing probe.

File: prototypes/mu_cosine/REPORT_eval_methodology.md [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+**Net (final, after §4.6's full objective fix):** the symmetric-trained checkpoint lost to a trained e5-probe —
+but that was **entirely the objective** (symmetric-dominant training, then regression-to-constants), not the
+architecture. With the objective fixed — directional supervision + a **discriminative ranking loss** — **μ beats
+the trained-head-on-frozen-e5 probe on both direction (0.982 vs 0.930) and close-negatives (0.864 vs 0.790)**,
+held-out, with non-overlapping CIs — **and it holds under a node-disjoint, 3-seed control (0.961 vs 0.920;
+0.812 vs 0.777, §4.6).** So μ's nonlinear architecture **does add value over a linear probe**, once trained for
+the task; the earlier loss was supervision, not capacity. The review's control was the right test and it drove
```
Comment (s243a) [inline prototypes/mu_cosine/REPORT_eval_methodology.md]: **[nit] Scope the architecture claim to the probe that was actually tested**

The updated control is a strong and useful result, but the probe here is still a linear order-aware head. Please phrase the claim as “μ beats a linear e5-probe under the directional-rank objective” and keep a capacity-matched nonlinear probe / LoRA / fine-tune as future work, rather than implying all trained e5 adaptations are exhausted.

File: prototypes/mu_cosine/REPORT_eval_methodology.md [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+  not reproducible from committed tooling, are now produced by committed scripts with DAG sibling-filters and
+  bootstrap CIs (`eval_arch_control.py`) and a shared train/test split (`build_triples`); the directional loss is
+  in the main trainer (`--dir-rank-weight`). The architecture-control + node-disjoint multi-seed are
+  `eval_arch_control.py` / `eval_arch_harden.py`.
+- **Architecture control caveat (resolved).** e5-cos was a *product* baseline; the *architecture* control (a
+  trained head on frozen e5) is now run (§4.6) and μ beats it once trained for the task.
+- **Still open:** an e5-*probe* learning curve for §4.1 (the filing curve beat e5-*cos* only); probabilistic
+  calibration (ECE/reliability) is unmeasured (§5); the node-disjoint base model (`model_nodetype`) still saw the
+  held-out nodes during *its* pretraining (only the task-example supervision is held out — symmetric with e5's
+  pretraining, but not a from-scratch node-disjoint train).
```
Comment (s243a) [inline prototypes/mu_cosine/REPORT_eval_methodology.md]: **[nit] Align node-disjoint wording with the pretraining caveat**

This caveat is good and should be reflected wherever §4.6 says the node-disjoint comparison is symmetric. The task-example split is node-disjoint, but μ inherits `model_nodetype` pretraining exposure while the probe starts from frozen e5 features, so the safest wording is “node-disjoint task supervision,” not pretraining-disjoint.

File: prototypes/mu_cosine/train_mu_attention.py [hunk 11]
```diff
@@ -565,6 +565,15 @@ def bld(items, **kw):                                  # tokenize on CPU, move t
...
+    # JUDGE→LOSS ROUTING (DESIGN): graph-judged DIRECTIONAL edges are *certain + oriented*, so train them with a
+    # RANKING loss (μ(member|container) > μ(container|member)) — discriminative, not regression-to-0.90/0.10
+    # (which caps separation). haiku/soft rows keep regression (L_graded/L_blend). The judge token already on
+    # every row is the router; no new embedding needed. (Beats the e5-probe on direction — REPORT §4.6.)
+    _DIR_REL = {"element_of", "subcategory", "subtopic", "super_category"}
+    dir_edges = [r for r in graded_tr if r[8] == "graph" and r[4] in _DIR_REL and r[2] >= 0.5]
+    # when the directional RANKING loss is on, graph-directional rows are rank-supervised (+ anchor) — exclude them
+    # from the REGRESSION pool so the 0.90/0.10 regression doesn't compete with the rank (haiku/lateral rows stay).
+    graded_nondir = [r for r in graded_tr if not (r[8] == "graph" and r[4] in _DIR_REL)]
     relset = sorted(set(r[4] for r in graded_tag))
```
Comment (s243a) [inline prototypes/mu_cosine/train_mu_attention.py:576]: **[nit] Dead variable / stale comment**

`graded_nondir` is defined here, but the updated design intentionally sends directional graph rows through both regression and rank paths later. If this variable is no longer used, please delete it and the stale comment so future readers do not think directional rows are excluded from the regression pool.

**Review by s243a — COMMENTED** [review]
Follow-up council re-review after the `e72baefa` update: **approve with nits**. The prior request-changes blockers are substantively resolved. The directionality/sibling metrics now have committed tooling and node-disjoint 3-seed controls; the calibration wording is corrected to dynamic range/readability; threats-to-validity are much more complete; baseline split drift is explained; and the trained e5-probe control materially improves the methodology.

Remaining items are non-blocking cleanup: reconcile a stale filing-baseline sentence, scope the architectural-superiority prose to the actual linear-probe control, and remove a dead/stale variable in the trainer. The Haiku lateral layer is appropriately framed as a 17-pair proof-of-pipeline rather than a full-scale result.
